From de23aca73f3a5aa933a35efba818142e7033ccbd Mon Sep 17 00:00:00 2001
From: Roman Stratiienko <r.stratiienko@gmail.com>
Date: Thu, 19 Oct 2023 03:37:01 +0300
Subject: [PATCH 1/3] HWUI: Fix deadlock in graphics pipeline

Mutex lock() causes deadlock when the queue is empty.
Regression is caused by b013c8dee3b3e897b3557206f2d6c75b8c9f2f84
For some unknown reason, it is reproduced only on GloDroid :)
So probably some other part of GD is to blame.
Or, it is just a coincidence.

TODO: Do one more investigation attempt.

Change-Id: Ic981947babe62b83032e9ba1630cc0458c228f44
Signed-off-by: Roman Stratiienko <r.stratiienko@gmail.com>
---
 libs/hwui/renderthread/CanvasContext.cpp | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/libs/hwui/renderthread/CanvasContext.cpp b/libs/hwui/renderthread/CanvasContext.cpp
index 504dfaa2a1f5..818add86bff9 100644
--- a/libs/hwui/renderthread/CanvasContext.cpp
+++ b/libs/hwui/renderthread/CanvasContext.cpp
@@ -587,7 +587,11 @@ void CanvasContext::draw(bool solelyTextureViewUpdates) {
         // not visible to IRenderPipeline much less FrameInfoVisualizer. And since this is
         // the thread we're primarily concerned about being responsive, this being too broad
         // shouldn't pose a performance issue.
-        std::scoped_lock lock(mFrameMetricsReporterMutex);
+
+// Causes deadlock when the queue is empty. Regression is caused by b013c8dee3b3e897b3557206f2d6c75b8c9f2f84
+// For some unknown reason, it is reproduced only on GloDroid :) So probably some other part of GD is to blame. Or, it is just a coincidence.
+//        std::scoped_lock lock(mFrameMetricsReporterMutex);
+
         drawResult = mRenderPipeline->draw(frame, windowDirty, dirty, mLightGeometry,
                                            &mLayerUpdateQueue, mContentDrawBounds, mOpaque,
                                            mLightInfo, mRenderNodes, &(profiler()), mBufferParams);
-- 
2.34.1

